docs/api.rst

```diff
@@ -0,0 +1,287 @@
+      ``a`` to an empty dict, then proceeds.
+
+      >>> tree_set({}, 'a.b', value=5, create_missing=True)
+      {'a': {'b': 5}}
+
+      This tries to traverse ``a``, but it doesn't exist and it's not the last
+      edge in the path. The next edge is ``[2]``, which is an index, so it first
+      sets ``a`` to a list of 3 ``None`` values, then proceeds.
+
+      >>> tree_set({}, 'a.[2]', value=5, create_missing=True)
+      {'a': [None, None, 5]}
+
+      This is similar, but with a negative index.
+
+      >>> tree_set({}, 'a.[-1]', value=5, create_missing=True)
+      {'a': [5]}
+
+      This creates missing indices in an existing list.
+
+      >>> tree_set({'a': []}, 'a.[2]', value=5, create_missing=True)
+      {'a': [None, None, 5]}
+
+
+   Examples:
+
+   These don't mutate the tree:
+
+   >>> tree = {'a': {'b': {'c': 1}}}
+   >>> tree_set(tree, 'a', value=5, mutate=False)
+   {'a': 5}
+   >>> tree_set(tree, 'a.b.c', value=[], mutate=False)
+   {'a': {'b': {'c': []}}}
+
+   These raise errors if an edge is missing:
+
+   >>> tree_set({}, 'a.b.c', value=5)
+   KeyError ...
+   >>> tree_set({}, 'a.[1].b', value=5)
+   IndexError ...
+
+   These create missing edges and indexes:
+
+   >>> tree_set({}, 'a.b.c', value=5, create_missing=True)
+   {'a': {'b': {'c': 5}}}
+   >>> tree_set({}, 'a.[1].b', value=5, create_missing=True)
+   {'a': [None, {'b': 5}]}
+
+
+.. py:func:: tree_flatten(tree)
+
+   Flattens a tree into a dict with keys of paths.
+
+   >>> tree_flatten({'a': 1})
+   {'a': 1}
+   >>> tree_flatten({'a': {'b': 1, 'c': 2}})
+   {'a.b': 1, 'a.c': 2}
+   >>> tree_flatten({'a': [{'b': 1}, {'c': 2}]})
+   {'a.[0].b': 1, 'a.[1].c': 2}
+
+   .. Note::
+
+      At this point, a flattened tree can't be used using ``tree_get`` and
+      ``tree_set``.
+
+
+.. py:func:: tree_setdefault(tree, default_tree)
+
+   FIXME
+
+
+.. py:func:: tree_validate(tree, schema)
+
+   FIXME
+
+
+.. py:func:: tree_traverse(tree, fun)
+
+   FIXME
+
+
+Research and Inspirations
+=========================
+
+Python ``defaultdict``
+----------------------
+
+Python has a defaultdict
+
+https://docs.python.org/3/library/collections.html#defaultdict-objects
+
+This doens't handle lists and dicts well, though.
+
+We'd have to either create the original data structure as a defaultdict, or
+convert it to one.
+
+If you try to get something deep from a defaultdict, it mutates the
+structure.
+
+It doesn't easily support composable paths.
+
+
+jq processor
+------------
+
+jq has interesting filter syntax.
+
+https://stedolan.github.io/jq/manual/#Basicfilters
+
+
+Creating a new subclass of Python ``dict``
+------------------------------------------
+
+We could do that and add ``get_path`` and ``set_path``, but I wonder if we can
+get the utility we want without having to box/unbox data.
+
+If we're just working with dicts and lists and standard Python things, then
+``json.dumps`` and other things just work without us having to do anything about
```

Comment:
Are you familiar with custom `json.dumps` and `loads` decoders and encoders? You can arrange it so you get fancier, custom types out rather than vanilla dicts.